### PR TITLE
Reduce kernel status flickr, fewer affinity change

### LINKED ISSUE #11988: Refactor/Cleanup preferred controller and controller registration and related code
* Remove the following code in loadControllersImpl
```
// This is temporary, when we create an MRU list in VS Code or the like, this should go away.
```

### PR DESCRIPTION
**Pull request by DonJayamanne**
Fixes #11988

### CODE REVIEW

File: src/notebooks/controllers/controllerRegistration.ts [hunk 1]
```diff
@@ -82,26 +82,30 @@ export class ControllerRegistration implements IControllerRegistration {
...
     }
     addImpl(
         metadata: KernelConnectionMetadata,
         types: ('jupyter-notebook' | 'interactive')[],
         triggerChangeEvent: boolean
-    ): IVSCodeNotebookController[] {
-        let results: IVSCodeNotebookController[] = [];
+    ): { added: IVSCodeNotebookController[]; existing: IVSCodeNotebookController[] } {
+        const added: IVSCodeNotebookController[] = [];
+        const existing: IVSCodeNotebookController[] = [];
```
Comment (DonJayamanne) [inline src/notebooks/controllers/controllerRegistration.ts:108]: We were triggering change events even when we're updating existing items
This had other side effects, the change event only supports `added, removed`, not `updated`

File: src/notebooks/controllers/controllerRegistration.ts [hunk 2]
```diff
@@ -82,26 +82,30 @@ export class ControllerRegistration implements IControllerRegistration {
...
         metadatas.forEach((metadata) => {
-            added.push(...this.addImpl(metadata, types, false));
+            const { added } = this.addImpl(metadata, types, false);
+            addedList.push(...added);
         });
 
-        this.changeEmitter.fire({ added, removed: [] });
-        return added;
+        if (addedList.length) {
+            this.changeEmitter.fire({ added: addedList, removed: [] });
```
Comment (DonJayamanne) [inline src/notebooks/controllers/controllerRegistration.ts:92]: No more unnecessary change events, which had side effects

**Review by mjbvz — APPROVED** [review]
